### ISSUE TITLE
make clean_state and enforce_consistent_e higher-order friendly

### ISSUE THREAD

**Issue by zingale** [AMReX-Astro/Castro#290]
These routines that allow for higher-order schemes

**Comment by zingale** [comment]
actually, `enforce_consistent_e` is okay -- we only call that right after initialization, so as long as we call this before we convert to cell-averages, we are okay.

**Comment by zingale** [comment]
`computeTemp` and `reset_internal_e` need to be addressed, however.

**Comment by zingale** [comment]
I can confirm that if these 2 routines are commented out, then we get 4th order convergence.  Haven't looked to see if they are both problematic or if there is only one issue.

**Comment by zingale** [comment]
For `reset_internal_e`, we work directly off of `S_new`.  If we do the cell-average -> cell-center conversion in place, then we risk losing conservation, so it is best to copy the state into a buffer, like `Sborder` and do the conversion there.  

Only `UEINT` and `UEDEN` are updated here, so we only would need to copy those back into `S_new`, ideally, only for the zones that were updated.

But it may be that we do not need to worry about this routine at all.  Any of the reset stuff is going to break convergence by design, since they are not working from the PDEs anymore, but just enforcing certain constraints that we want.  So this means that convergence goes out the window anyway.  This argues that we should just leave `reset_internal_e` as is.

`computeTemp` should be updated though to give a 4th order accurate T.

**Comment by zingale** [comment]
This is actually tricker than I thought, since `clean_state` and `computeTemp` take a MF as an argument instead of working directly on StateData.  I need to do some ghost cell fills, so I need to put things into StateData at some point.

It might be cleaner if we rewrote `clean_state` and `computeTemp` to work from `StateData` directly...

**Comment by zingale** [comment]
or we create a new StateData called `SDC_Scratch` that is just used for ghost filling and is basically the same as `State_Type`

**Comment by zingale** [comment]
@maxpkatz thoughts welcomed here.

tl;dr: I need to convert between cell averages and cell-centers in `computeTemp()`, which means I need to fill ghost cells on the incoming `state`

**Comment by maxpkatz** [comment]
I am OK with rewriting computeTemp/clean_state to work on StateData. I think that is the right solution here. Then we can just do a ghost cell fill at the beginning of the call.

**Comment by maxpkatz** [comment]
The latter.

**Comment by zingale** [comment]
okay... here's a survey of where `clean_state` is currently called.

* `Castro_advance.cpp` : 
   this always operates on `StateData`, so we should be able to switch the interface w/o problems here.

* `Castro_react.cpp` : 
  there are 2 places here, the first half of Strang and the second half of Strang.  For the 2nd half,
  we are working on `S_new`, so there is no issue.  We can move `clean_state` out of this routine even,
  and call it in the driver.

  For the 1st half, we a working on `Sborder`.  After the burn, we copy into `S_new`, so it is conceivable 
  that we can do a `clean_state` on that and then copy back into `Sborder`.  Otherwise, we will need to 
  maintain a version of `clean_state` that operates on a MF.

  Note: the 4th order stuff won't do Strang, so this won't be an issue.

* `Castro_source.cpp` : 
  There are two places we call it.  In `apply_source_to_state`, we operate on a MF and then call `clean_state`.  We can simply move `clean_state` out of the calling routine that call it in the driver after the apply, on the `StateData`.  It appears that this is always done with `S_new`, so this should not be a problem.

  In the old-school SDC, in `get_react_source_prim`, we call `clean_state` on a MF.  This is a MF operation, and it is not clear how to generalize this to `StateData`, so this might mean keeping a MF version of `clean_state` around.  Again, this is not going to affect this issue, since 4th order code won't go down this path.

**Comment by zingale** [comment]
To start, I'll make a branch with a version of `clean_state` that works on `StateData` and move everything I can over to it, and test it.

**Comment by zingale** [comment]
one issue with expand_state is that if there are ghost cells in the `State_Type` `StateData`, then we need to be careful about rearranging the `expand_state` to ensure that we don't try to enforce minimum density on uninitialized ghost cells.

**Comment by zingale** [comment]
Note: this work is being done on `clean_state`

It was tested here:

http://groot.astro.sunysb.edu/Castro/test-suite/gfortran/2018-05-04-001/index.html

**Comment by zingale** [comment]
the previous test worked:

http://groot.astro.sunysb.edu/Castro/test-suite/gfortran/2018-05-03-002/index.html

the failures arise from the reorg of expand_state

**Comment by zingale** [comment]
the radiation failures are likely because when we do the clean_state first, it will try to operate on the ghost cells, which are nonsense.  So we need an option to tell `clean_state` to ignore ghost cells.

**Comment by zingale** [comment]
The `clean_state` branch not works, with small "errors" from the rearranging.  All of the needed routines now take a number of ghost cells so we can do a `clean_state()` call in `expand_state()` first, w/o ghost cells and then do the `FillPatch`.

The next trick is to make `computeTemp()` do the right thing.  We need to do some `expand_state()`s there, but we will want to do it wilhout a `clean_state()` to avoid circular logic.  We can pass `-1` in as the state slot to operate on to avoid the `clean_state()` in `expand_state()`

**Comment by zingale** [comment]
this now works on the `sdc` branch.

**Comment by zingale** [comment]
turns out we do need to worry about `reset_internal_e`, atleast according to the `acoustic_pulse_general` problem.  We converge better if we comment it out

**Comment by zingale** [comment]
the fix for this is to move `reset_internal_e` into the region where we converted from cell-averages to cell-centers.

**Comment by zingale** [comment]
this is now completed and the general EOS version converges 4th order
